### ISSUE TITLE
Fix discrepancies in `odo describe` machine readable and human readable output

### ISSUE THREAD

**Issue by valaparthvi** [redhat-developer/odo#5344]
There are discrepancies between the `odo decribe` output for a pushed component when run within the context directory(or with --context), and outside the context directory without the --context flag. We cannot clearly define the state of storage, and urls, they could be locally deleted, but available on the cluster, or vice versa without having access to both devfile(local) and in-cluster(remote) data. 

<details>
<summary>Outside the context directory:</summary>

```yaml
{
	"kind": "Component",
	"apiVersion": "odo.dev/v1alpha1",
	"metadata": {
		"name": "mynode2",
		"namespace": "default",
		"creationTimestamp": null,
		"labels": {
			"app": "app",
			"app.kubernetes.io/instance": "mynode2",
			"app.kubernetes.io/managed-by": "odo",
			"app.kubernetes.io/managed-by-version": "v2.4.3",
			"app.kubernetes.io/name": "",
			"app.kubernetes.io/part-of": "app",
			"component": "mynode2"
		},
		"annotations": {
			"deployment.kubernetes.io/revision": "1",
			"odo.dev/project-type": "nodejs"
		}
	},
	"spec": {
		"app": "app",
		"type": "nodejs",
		"urls": {
			"kind": "List",
			"apiVersion": "odo.dev/v1alpha1",
			"metadata": {},
			"items": [
				{
					"kind": "URL",
					"apiVersion": "odo.dev/v1alpha1",
					"metadata": {
						"name": "mynode2-3030",
						"creationTimestamp": null
					},
					"spec": {
						"host": "mynode2-3030.192.168.59.101",
						"protocol": "http",
						"port": 3030,
						"secure": false,
						"kind": "ingress",
						"path": "/"
					},
					"status": {
						"state": ""
					}
				}
			]
		},
		"storages": {
			"metadata": {},
			"items": [
				{
					"kind": "Storage",
					"apiVersion": "odo.dev/v1alpha1",
					"metadata": {
						"name": "store1",
						"creationTimestamp": null
					},
					"spec": {
						"size": "1Gi",
						"path": "/data",
						"containerName": "runtime"
					}
				}
			]
		},
		"env": [
			{
				"name": "PROJECTS_ROOT",
				"value": "/project"
			},
			{
				"name": "PROJECT_SOURCE",
				"value": "/project"
			},
			{
				"name": "DEBUG_PORT",
				"value": "5858"
			}
		]
	},
	"status": {
		"state": "Pushed"
	}
}
```

</details>

<details>
<summary>Inside the context directory:</summary>

```yaml
{
	"kind": "Component",
	"apiVersion": "odo.dev/v1alpha1",
	"metadata": {
		"name": "mynode2",
		"namespace": "default",
		"creationTimestamp": null,
		"labels": {
			"app": "app",
			"app.kubernetes.io/instance": "mynode2",
			"app.kubernetes.io/managed-by": "odo",
			"app.kubernetes.io/managed-by-version": "v2.4.3",
			"app.kubernetes.io/name": "",
			"app.kubernetes.io/part-of": "app",
			"component": "mynode2"
		},
		"annotations": {
			"deployment.kubernetes.io/revision": "1",
			"odo.dev/project-type": "nodejs"
		}
	},
	"spec": {
		"app": "app",
		"type": "nodejs",
		"urls": {
			"kind": "List",
			"apiVersion": "odo.dev/v1alpha1",
			"metadata": {},
			"items": [
				{
					"kind": "URL",
					"apiVersion": "odo.dev/v1alpha1",
					"metadata": {
						"name": "mynode2-3030",
						"creationTimestamp": null
					},
					"spec": {
						"host": "mynode2-3030.192.168.59.101",
						"protocol": "http",
						"port": 3030,
						"secure": false,
						"kind": "ingress",
						"path": "/"
					},
					"status": {
						"state": "Pushed"
					}
				}
			]
		},
		"storages": {
			"kind": "List",
			"apiVersion": "odo.dev/v1alpha1",
			"metadata": {},
			"items": [
				{
					"kind": "Storage",
					"apiVersion": "odo.dev/v1alpha1",
					"metadata": {
						"name": "store1",
						"creationTimestamp": null
					},
					"spec": {
						"size": "1Gi",
						"path": "/data",
						"containerName": "runtime"
					},
					"status": "Pushed"
				}
			]
		},
		"env": [
			{
				"name": "PROJECTS_ROOT",
				"value": "/project"
			},
			{
				"name": "PROJECT_SOURCE",
				"value": "/project"
			},
			{
				"name": "DEBUG_PORT",
				"value": "5858"
			}
		]
	},
	"status": {
		"state": "Pushed"
	}
}
```

</details>

_Originally posted by @valaparthvi in https://github.com/redhat-developer/odo/issues/5245#issuecomment-1001400170_

Apart from this, there are 2 structs at play here, [ComponentFullDescriptionSpec](https://github.com/redhat-developer/odo/blob/main/pkg/component/component_full_description.go#L22) vs [ComponentSpec](https://github.com/redhat-developer/odo/blob/main/pkg/component/types.go#L22), which are almost similar, except for a few fields.

**Comment by dharmit** [comment]
I think we concluded a discussion on this as "expected" behaviour, no? I'm just wondering if there's something to be fixed in the code or added into the docs. 🤔 



> Apart from this, there are 2 structs at play here, [ComponentFullDescriptionSpec](https://github.com/redhat-developer/odo/blob/main/pkg/component/component_full_description.go#L22) vs [ComponentSpec](https://github.com/redhat-developer/odo/blob/main/pkg/component/types.go#L22), which are almost similar, except for a few fields.

`ComponentSpec` has more fields than `ComponentFullDescriptionSpec` even though the latter has "full" in its name. 😀 We should remove one of them.

IMO, we should also move all type definitions from `pkg/component/component_full_description.go` to `pkg/component/types.go` because that's what `types.go` file usually is meant for.

**Comment by valaparthvi** [comment]
@kadel If the localconfig is accessible, but the cluster is not, or vice versa (cluster is accessible, but localconfig is not), should we show the state of Storage and URL as Unknown?

**Comment by kadel** [comment]
I would use `unknown` only if the cluster is not accessible. If you can check cluster but you don't have local info from devfile.yaml it should show it as Pushed

**Comment by valaparthvi** [comment]
tldr; `odo describe` should return the information that user wants, instead of returning the information that odo has.

There are a few other things as well.
Take for example:

 `abc` is a directory representing component `ABC`, within this context directory, we run odo describe

1. User runs `odo describe GEG --app app --project default` for a pushed component GEG, then instead of returning information about component GEG, odo returns the information about component ABC
2. User runs `odo describe SMH --app --project default` for a non-pushed component, then again, instead of returning information about component ABC, it should return whatever information odo has about SMH, i.e. component name, app name, and project name. This behaviour is similar to running `odo describe` for a component outside a context directory.

If `odo describe` is supposed to provide detailed information about the components that you see in `odo list`, then it should not be bound to a directory and it's component, odo should return information about the component that the user has requested.

I have created a POC(#5370) of how I think `odo describe` should behave, perhaps you can test and see if it makes sense.

**Comment by dharmit** [comment]
> 1. User runs `odo describe GEG --app app --project default` for a pushed component GEG, then instead of returning information about component GEG, odo returns the information about component ABC

Is it just me who finds this opposite of the tl;dr - `odo describe should return the information that user wants, instead of returning the information that odo has.`?

If odo was indeed to provide the info that user wants, it should return information about GEG instead of ABC in above case. Or did you mistakenly write the reverse, because you seem to have summarized according to tl;dr for component SMH.

> If `odo describe` is supposed to provide detailed information about the components that you see in `odo list`, then it should not be bound to a directory and it's component, odo should return information about the component that the user has requested.

But how? For detailed information, we do need the devfile and env.yaml, no? How are we planning to implement this if `odo describe` is executed outside a component directory?

**Comment by valaparthvi** [comment]
> > 1. User runs `odo describe GEG --app app --project default` for a pushed component GEG, then instead of returning information about component GEG, odo returns the information about component ABC
> 
> Is it just me who finds this opposite of the tl;dr - `odo describe should return the information that user wants, instead of returning the information that odo has.`?
> 
> If odo was indeed to provide the info that user wants, it should return information about GEG instead of ABC in above case. Or did you mistakenly write the reverse, because you seem to have summarized according to tl;dr for component SMH.

I meant the current behaviour of odo is such that it return information about component from the current directory, instead of the one that user asked.

> > If `odo describe` is supposed to provide detailed information about the components that you see in `odo list`, then it should not be bound to a directory and it's component, odo should return information about the component that the user has requested.
> 
> But how? For detailed information, we do need the devfile and env.yaml, no? How are we planning to implement this if `odo describe` is executed outside a component directory?

If the component is pushed, and you do not have access to the devfile and env.yaml, you simply return the pushed component, but if it is not a pushed component, you simply return it's name, app name, and project name as was passed by the user.

**Comment by kadel** [comment]
> But how? For detailed information, we do need the devfile and env.yaml, no? How are we planning to implement this if `odo describe` is executed outside a component directory?

For the majority of the information about components, you don't need env.yaml. You can get all information from cluster or from devfile.yaml. 

If you don't have env.yaml but you know the component name, application and project you can get all details about the component from the cluster.

**Comment by dharmit** [comment]
> For the majority of the information about components, you don't need env.yaml. You can get all information from cluster or from devfile.yaml.

I don't find the idea of presenting incomplete information to be very useful. I'm looking at this as a zero or one problem, in that you either give the user full info or no info. The direction this issue is taking is about having a middle ground when context is not available. And it's fine if we have decided to have that kind of UX. It's just something I personally am not in a favor of.

**Comment by kadel** [comment]
> I don't find the idea of presenting incomplete information to be very useful. I'm looking at this as a zero or one problem, in that you either give the user full info or no info.

What information would be missing?

**Comment by dharmit** [comment]
Storage and URL. Their state won't be accurate when the user does `odo describe` from outside a component directory, and there's no way for odo to be accurate about it.

**Comment by kadel** [comment]
> Storage and URL. Their state won't be accurate when the user does `odo describe` from outside a component directory, and there's no way for odo to be accurate about it.

We already show `unknown` state when odo can read local devfile and `env.yaml` but can't reach the cluster. Why not make it consistent and show `unknown` when we can read info from the cluster but not from local files?

**Comment by dharmit** [comment]
> Why not make it consistent and show `unknown` when we can read info from the cluster but not from local files?

In a previous comment on this issue(https://github.com/redhat-developer/odo/issues/5344#issuecomment-1017665708), you asked to use `Pushed` for such a case. It gets confusing pretty quickly as to what "exactly" the states `Unknown`, `Pushed` etc. mean, and even we (the contributors) would have a tough time remembering at times. This is why I would like for things to be zero or one, as long as possible. Is `odo describe --app <blah> --project <more-blah>` worth an exception?

**Comment by kadel** [comment]
> > Why not make it consistent and show `unknown` when we can read info from the cluster but not from local files?
> 
> In a previous comment on this issue([#5344 (comment)](https://github.com/redhat-developer/odo/issues/5344#issuecomment-1017665708)), you asked to use `Pushed` for such a case.

Sorry I had a side discussion with @valaparthvi. I admitted that what she was proposing is better (showing `Unknown` in this case).

**Comment by dharmit** [comment]
> Sorry I had a side discussion with @valaparthvi. I admitted that what she was proposing is better (showing `Unknown` in this case).

OK. 👍🏾 

Can we please also write docs for this issue? Mainly command reference docs for `odo list` and `odo describe` commands.
